### DISCUSSION TITLE
RAG as a Tool

### DISCUSSION THREAD

**Discussion by LizeRaes** [langchain4j/langchain4j#1934] [Ideas]
What do you think of the idea of extending the infrastructure to make it easy to add RAG as a Tool?
Is this already feasible / easy today?

Something like this
```
@Tool("Retrieve internal documents, database info and web search info that are relevant to the user question. The query needs to contain all relevant information and keywords in order to find all relevant context.")
retrieveRelevantContext(String query){}
```

I see the following advantages of having RAG as a Tool:
- the model can request context iteratively (am I correct that iterative searching in RAG is not supported today?). An example use case I had was: based on a user-picked disease, the LLM finds a protein name in some pdfs (EmbeddingStoreContentRetriever), and sequence of that protein from a database (SqlContentRetriever). This has to happen in two steps, because only when the protein name is found, can the correct sql query by generated. I ended up passing a tool to query the db, with the schema in the tool description, as a functional workaround.
- the model will only retrieve context when it thinks it needs it, which saves tokens and processing power for
-- all the cases where the previously pulled content was sufficient already
-- all the small interactions where the context is unimportant

Curious to hear your take on this idea!

**Comment by mczcqqchen** [comment]
@langchain4j May I ask if Tools supports Llama3.2 or qwen2?

**Comment by mczcqqchen** [comment]
Thank you




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "langchain4j/langchain4j"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月17日(星期四) 下午3:29
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [langchain4j/langchain4j] RAG as a Tool (Discussion #1934)





 
@mczcqqchen you can see which providers support tools here: https://docs.langchain4j.dev/integrations/language-models/
 
I guess you can use Llama3.2 or qwen2 with Ollama, they should support tools. But the quality of course will not be comparable to GPT-4o or claude3.5
